xdsl/transforms/eqsat_optimize_pdl_interp.py

```diff
@@ -0,0 +1,155 @@
+from dataclasses import dataclass, field
+
+from xdsl.context import Context
+from xdsl.dialects import builtin, pdl_interp
+from xdsl.ir import Block, SSAValue, Use
+from xdsl.passes import ModulePass
+from xdsl.traits import SymbolTable
+
+
+@dataclass(frozen=True)
+class EqsatOptimizePDLInterp(ModulePass):
+    name = "eqsat-optimize-pdl-interp"
+
+    def apply(self, ctx: Context, op: builtin.ModuleOp) -> None:
+        matcher = SymbolTable.lookup_symbol(op, "matcher")
+        assert isinstance(matcher, pdl_interp.FuncOp)
+        assert matcher is not None, "matcher function not found"
+
+        mo = MatcherOptimizer(matcher)
+        mo.optimize()
+
+
+@dataclass
+class MatcherOptimizer:
+    matcher: pdl_interp.FuncOp
+
+    finalize_blocks: set[Block] = field(default_factory=set[Block])
+
+    def optimize(self):
+        for op in self.matcher.walk(reverse=True):
+            if isinstance(op, pdl_interp.AreEqualOp):
+                self.optimize_equality_constraints(op)
+
+    def optimize_equality_constraints(
+        self,
+        are_equal_op: pdl_interp.AreEqualOp,
+    ) -> bool:
+        if not self.is_finalize_block(are_equal_op.false_dest):
+            # If the false destination is not a finalize block, we cannot optimize this AreEqualOp.
+            return False
+
+        assert (are_equal_block := are_equal_op.parent_block())
+        blocks_to_move = [are_equal_block]
+
+        # We can always pick only one of the predecessors to move up.
+        # Forked paths will never define values that are used after paths
+        # are joined again, since there are no block arguments/phi nodes.
+        # The final destination of the blocks will always be along a
+        # non-split path at the post-dominator frontier:
+        assert (incoming_edge := next(iter(are_equal_block.uses), None)) is not None
+        incoming_edges: list[Use] = [incoming_edge]
+
+        outgoing_edges = [are_equal_op._successor_uses[0]]  # pyright: ignore[reportPrivateUsage]
```

Comment:
Why use the private API here?

Reply:
hmm, indeed, maybe actually we just need to expose this as public API? CC @math-fehr